tests/pos/suspend-strawman-2/monadic-reflect.scala

```diff
@@ -0,0 +1,56 @@
+import scala.util.boundary
+import runtime.*
+
+trait Monad[F[_]]:
+
+  /** The unit value for a monad */
+  def pure[A](x: A): F[A]
+
+  extension [A](x: F[A])
+    /** The fundamental composition operation */
+    def flatMap[B](f: A => F[B]): F[B]
+
+    /** The `map` operation can now be defined in terms of `flatMap` */
+    def map[B](f: A => B) = x.flatMap(f.andThen(pure))
+
+end Monad
+
+trait CanReflect[M[_]]:
+  def reflect[R](mr: M[R]): R
+
+trait Monadic[M[_]: Monad]:
+
+  /**
+   * Embedding of pure values into the monad M
+   */
+  def pure[A](a: A): M[A]
+
+  /**
+   * Sequencing of monadic values
+   *
+   * Implementations are required to implement sequencing in a stack-safe
+   * way, that is they either need to implement trampolining on their own
+   * or implement `sequence` as a tail recursive function.
+   *
+   * Actually the type X can be different for every call to f...
+   * It is a type aligned sequence, but for simplicity we do not enforce this
+   * here.
+   */
+  def sequence[X, R](init: M[X])(f: X => Either[M[X], M[R]]): M[R]
+
+  /**
+   * Helper to summon and use an instance of CanReflect[M]
+   */
+  def reflect[R](mr: M[R])(using r: CanReflect[M]): R = r.reflect(mr)
+
+  /**
+   * Reify a computation into a monadic value
+   */
+  def reify[R](prog: CanReflect[M] ?=> R): M[R] =
+    boundary [M[R]]:
+      given CanReflect[M] with
+        def reflect[R2](mr: M[R2]): R2 =
+          suspend [R2, M[R]] (s => mr.flatMap(s.resume))
```

Comment:
Thanks! I was just about to propose this. I mixed up that for my project loom experiments the casts were necessary since they implement coroutines and not proper continuations.

Reply:
Yes, indeed! I am very pleased how everything simply holds together.

Reply:
In retrospect, shouldn't be too surprising recalling Filinski's original equation:

![image](https://user-images.githubusercontent.com/408265/216618328-5e4a20b1-487f-4c49-92bf-5f8410b85e2f.png)

which in Scalaish syntax translates to:

```scala
def reify(e) = boundary(pure(e))
def reflect(e) = suspend(k => e.flatMap(k.resume))
```

`*` is bind in his paper:
![image](https://user-images.githubusercontent.com/408265/216618697-e3cbf7e7-bdbf-4641-aa03-118464b92d42.png)

Reply:
Neat! So this reproduction of Filinski's work shows that boundary/suspend can serve as the canonical way to reflect monads, and therefore to represent effects. That strengthens the case for it considerably.

Reply:
Yes! `boundary` is just a more friendly name for a (multiprompt) `reset` and `suspend` is a more friendly name for a multiprompt version of `shift0` :)

Reply:
@odersky @b-studios In response to how to suspend:

I apologize for the length of this ahead of time. The longer code examples are linked with gists below. The full examples and transformed outputs will be linked at the bottom of this reply.

We've completed our first round of the suspend CPS compiler transformations inline with the design we proposed in [Pre-SIP: Suspended functions and continuations in Scala 3.
](https://contributors.scala-lang.org/t/pre-sip-suspended-functions-and-continuations/5801). In brief, ours are one-shot delimited continuations. We derive the suspension point labels by analyzing the bodies of definitions declaring a `given` `Suspend` parameter. An example below:

```scala
package examples

import continuations.*

@main def ThreeDependentContinuations =
  def threeDependentContinuations(a: Int, b: Int, c: Int)(using s: Suspend): Int =
    val d = 4
    val continuationOne:Int = s.suspendContinuation(_.resume(Right(d + a)))
    val e = 5
    val continuationTwo:Int = s.suspendContinuation(_.resume(Right(continuationOne + e + b)))
    val f = 6
    val result:Int = s.suspendContinuation(_.resume(Right(continuationTwo + f + c)))
    result
  println(threeDependentContinuations(1,2,3))
```
https://gist.github.com/jackcviers/59c5d06c33a0a2508bec30a8f4c280a0

The output of the above program is:

```
21
```

During the transformation, a class extending the targeted platform's `ContinuationImpl` and the input variables and intermediary suspended results is generated. It implements `invokeSuspend` to call the user's transformed suspended definition. [Gist](https://gist.github.com/jackcviers/5642d3c8a2fa46b39b3b6e38069cfc93)

We then instantiate the generated continuation class and cycle through labels in the order of the resumed continuations, assigning the result to a local result variable, and jumping to the break label of the next continuation or dependent calculation. Each resume call is handeled by a `SafeContinuation` class instance, that stores the continuation state (Undecided, Suspended, Resumed, Error|Result) and ensures that the continuation cannot be resumed with more than one value. This state variable is a volatile variable, and we transition between states using compare and swap. The suspend parameter is removed from the declaration and a completion continuation parameter is inserted and the return type is modified to include `Null`:

```scala
def threeDependentContinuations(a: Int, b: Int, c: Int, 
  completion: continuations.Continuation[Int]
): Int | Null | (
  continuations.Continuation.State.Suspended : 
    continuations.Continuation.State
)
```

The state cases are rather repetitive, so I'll just paste a single case snippet of the user's transformed method in [Gist](https://gist.github.com/jackcviers/81c43082ade20684cde089bbed38224f).

This continues until we have a final value:

```scala
case 3 =>
  // ...
  result = $result.asInstanceOf[Int]
  // ...
```

Which is then returned.

It works with asynchronous resumes. However, due to the result returning the continuation state (`Suspended`) when resume has not yet been called, you will get `Suspended` as the  result of the CPS transformation:

```scala
...
given ExecutorService = Executors.newWorkStealingPool()

@main def ListMapNoResume = 
  def twoArgumentsOneContinuationsCFBefore(x: Int, y: Int): Suspend ?=> Int | Continuation.State.Suspended.type =
    println("twoArgumentsOneContinuationsCFBefore")
    val z = 1
    summon[Suspend].suspendContinuation[Int]{ c =>
      summon[ExecutorService].submit(new Runnable{
        override def run =
          val sleepTime = Random.between(10, 5000)
          println(s"sleepTime $sleepTime")
          Thread.sleep(sleepTime)
          c.resume(Right(x + y + z))})
    }

  val mappedContinuations = (1 to 100).toList.map(twoArgumentsOneContinuationsCFBefore(1, _))
  println(mappedContinuations)
```
Results [Gist](https://gist.github.com/jackcviers/f63d69c8465d357d8306379510821be9)

It is expected programmers would use `async/await` to block and wait until each continuation is resumed. Note this is current WIP.

No modification is made to List#map in this transformation:

```scala
val mappedContinuations: List[Matchable] = 
[info]           intWrapper(1).to(100).toList.map[Matchable](
[info]             {
[info]               def $anonfun(_$1: Int): Matchable = 
[info]                 twoArgumentsOneContinuationsCFBefore(1, _$1, 
[info]                   continuations.jvm.internal.ContinuationStub.contImpl
[info]                 )
[info]               closure($anonfun)
[info]             }
[info]           )
```

Since we are writing the scheduler and runtime, currently, we stub in an identity continuation for the initial continuation when calling suspended definitions to kick everything off at runtime. This, as with the interceptors, is planned to be platform specific.

There are some optimizations that we implement as well. For example, when a suspended definition has a constant body, we don't impose the use of a state machine:

```scala
def zeroArgumentsZeroContinuations()(using Suspend): Int = 1
//becomes
def zeroArgumentsZeroContinuations(
  completion: continuations.Continuation[Int | Any]
): Any = 1
```

Our `Suspend` is only a marker trait. Our continuation doesn't abstract over the input. We provide for a context tuple holding services that we can search by type to apply for contextual handlers, but I anticipate that being lifted to `givens` and/`capabilities` for things not necessary for different `initial continuations`. To make the interface as simple as possible to use we don't rely upon declared delimiters or additional parameters, though it does impose a cost in the implementation complexity.

I believe we can accommodate the strawmen here, though it will be different supporting multiple resumes, and puts the control of label insertion in the hands of the users. That may make the CPS transformation simpler, IMHO.

My motivation for the PRE-SIP was also Filinski 1994, and algebraic effects. I want interop with existing monadic libraries, but also to be able to compose multiple effects (State, Nondeterminism (choice in this proposal), error handling, console IO, monadic composition) without transformer stacks and eval trampolines for stack-unsafe monadic effects. Mostly things that are shown in[Matija Pretnar/2015/An Introduction to Algebraic Effects and Handlers](https://www.sciencedirect.com/science/article/pii/S1571066115000705).

Full source code of the above examples, and bytecode outputs available in this [Gist](https://gist.github.com/jackcviers/04efea109a5459a989533a6b10b72d2b). More working examples are available in the [47Deg/TBD repository](https://github.com/47deg/TBD/tree/main/continuationsPluginExample/src/main/scala/examples).

Reply:
Thanks for the pointer! I have some questions that I try to get ask for every scheme that tries to implement continuations: 

 - Which code is transformed to allow suspensions? I.e. what is the precise criterion that determines whether you transform code or leave it alone?
 - How is effect polymorphism handled? As an example, can you still work with a single definition of `map` with the standard signature? And is its code transformed or not?
 - How is the code transformed? CPS transform, state machines, capture/replay via exceptions, or something else?
 - What is the performance degradation for transformed code vs direct style code if the code does in the end not suspend at runtime? And is it the same after the first suspension?

Reply:
@odersky We've changed the definitions to match some of the Strawman
and literature, and we're in the process of making it match the strawman api. There is one exception in ours -- resume must be unit returning in order for the boundaries to work:

`shift` -> `suspend`

> Which code is transformed to allow suspensions
And
> How is effect polymorphism handled? As an example, can you still work with a single definition of map with the standard signature? And is its code transformed or not?


A suspended definition is a `ValOrDefDef` that includes `Suspend` in a
`ContextFunction` parameter position or in a `using`
`ParameterClause`.

E.g.

```scala
val suspendedFoo: (Suspend) => Int = ...
def suspendedFooDef(using Suspend): Int = ...
```

Once a `ValOrDefDef` has been determined to be a suspended definition,
if the RHS has no calls to `Suspend#shift`, then no transformation of
the RHS occurs:

```scala
def constantSuspendedDefinition(i: Int)(using Suspend): Int = 1
//becomes
def constantSuspendedDefinition(i: Int, completion: Continuation[Int]): Int = 1
```

If only a single call to `Suspend.shift` occurs, then we install the
`SafeContinuation`, which handles interception from the boundary
completion continuation, allowing for scheduling, cancellation, and
deferred completion. Otherwise, the code is normal:

```scala
def suspendsOneTime(i: Int)(using s:Suspend): Int = s.shift(continuation => continuation.resume(i + 1))
//becomes
def suspendsOneTime(i: Int, completion: Continuation[Int]): Any | Null | 
  (continuations.Continuation.State.Suspended : continuations.Continuation.State) = 
  val continuation1: continuations.Continuation[String] = completion
  val safeContinuation: continuations.SafeContinuation[String] =
    continuations.SafeContinuation.init[String](continuation1)
  {
    {
      safeContinuation.resume(i.+(1))
    }
  }
  safeContinuation.getOrThrow()
```

With two or more `shifts`, we generate a `Frame` class to hold the
inputs, intermediate state, `Frame` "program counter", and invoke and
interception machinery. Then we modify the suspended definition into a
state machine, switching on the `Frame` program counter, inserting
`CaseDef` clauses for each program counter in the suspended
definition, with the same `SafeContinuation` calls as above:

```scala
def threeDependentContinuations(a: Int, b: Int, c: Int)(using s: Suspend): Int =
    val d = 4
    val continuationOne: Int = s.shift(_.resume(d + a))
    val e = 5
    val continuationTwo: Int = s.shift(_.resume(continuationOne + e + b))
    val f = 6
    val result: Int = s.shift(_.resume(continuationTwo + f + c))
    result
```

becomes
[this](https://gist.github.com/jackcviers/856c81cb46df028164777e7462638efc#file-threedependentcontinuationscompilertransformation-scala)

Finally, calls to suspended definitions in source code are transformed
to include the boundary continuation (THIS IS CURRENT WIP AND IS
SUBJECT TO SMALL CHANGES):

```scala
blockingBoundary.createContinuation{
  threeDepenedentContinuations(1,2,3)
}

//becomes
val blockingBoundary = new BlockingBoundaryContinuation{}
threeDependentContinuations(1,2,3, blockingBoundary)
```

```scala
def suspendedDefCallsOtherSuspendedDef()(using Suspend): Int = otherSuspendedDef()
// becomes
def suspendedDefCallsOtherSuspendedDef(completion: Continuation[Int]): Int =
  otherSuspendedDef(completion)
```

The only problem comes with HOFs. For that we need `async` and `await`
and `awaitAll` for collections, or we need to reflect and reify other
monads. These can be introduced as a library:

```scala
def mapExample()(using Suspend): List[Int] =
  List(1,2,3).map(async{someSuspendedDef(_)).awaitAll // async returns Deferred[A], awaitAll waits until the deferreds all have resumed, or cancelled, an error in one cancels the others
```

We've also discussed the possibility of instrumenting all functions
and switching between the slow-path functions that actually suspend
and the uninstrumented version of the function at runtime with
@b-studios.
 
> How is the code transformed? CPS transform, state machines, capture/replay via exceptions, or something else?

The code is CPS transformed implemented as a compiler phase. We use a
Frame stack machine per suspended definition with a frame
counter-based switch table, goto labels, and return from label.

Boundaries are implemented as continuation objects and provide
interception, scheduling, and cancellation. The default boundary is
non-blocking -- that is if resume isn't called yet when control is
returned to the calling function, the suspension point can return
Suspended instead of the result type.

Boundaries can also be blocking (automatic await behavior), or
concurrent (return Deffered[A] results that have a blocking await
method to retrieve the result or throw if the suspension was resumed
with an error). Blocking and concurrent boundaries implement
structured concurrency with cancellation.

See above with the exact details.

We're willing to discuss and experiment with other transformations --
we've discussed some with @b-studios.

> What is the performance degradation for transformed code vs direct style code if the code does in the end not suspend at runtime?
>  And is it the same after the first suspension?


I have run jmh benchmarks with:

```
[info] # JMH version: 1.32
[info] # VM version: JDK 19, OpenJDK 64-Bit Server VM, 19+36-2238
[info] # VM options: <none>
[info] # Blackhole mode: full + dont-inline hint
[info] # Warmup: 5 iterations, 10 s each
[info] # Measurement: 5 iterations, 10 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 1 thread, will synchronize iterations
[info] # Benchmark mode: Throughput, ops/time
``` 


So, overhead of suspension vs raw scala code, and overhead of raw
scala code called after suspension

```scala
def SuspendedScalaCode =
  def suspendedScalaCode(a: Int, b: Int, c: Int)(using s: Suspend): Int =
    s.shift(_.resume(a + b + c))
  suspendedScalaCode(1,2,3)

def PassthroughAfterFirstSuspension =
  def passthroughAfterFirstSuspension(a: Int, b: Int, c: Int)(using s:Suspend): Int =
    val x = s.shift[Int](_.resume(a + b))
    x + c
  passthroughAfterFirstSuspension(1,2,3)

def Passthrough =
  def passthrough(a: Int, b: Int, c: Int)(using Suspend): Int =
    a + b + c
  passthrough(1, 2, 3)

def MeasureRawScala =
  def rawScala(a:Int, b: Int, c: Int): Int = a + b + c
  rawScala(1,2,3)
```

We determine the overhead of scala code after suspend by measuring the
overhead of SuspendedScalaCode and RawScala, then
PassthroughAfterFirstSuspension and RawScala, then measuring the
difference.

| Overhead Type                                 | Percentage Overhead |
|:----------------------------------------------|--------------------:|
|suspended without suspend overhead		| 0.47% |
|suspension overhead				| 92.79%  |
|regular scala code after suspension overhead	| 2.08% |